graphdatascience/graph/graph_cypher_runner.py

```diff
@@ -0,0 +1,134 @@
+import re
+from itertools import chain, zip_longest
+from typing import Any, Dict, Optional, Tuple
+
+from pandas import Series
+
+from ..query_runner.arrow_query_runner import ArrowQueryRunner
+from ..query_runner.query_runner import QueryRunner
+from ..server_version.server_version import ServerVersion
+from .graph_object import Graph
+from graphdatascience.caller_base import CallerBase
+
+
+class GraphCypherRunner(CallerBase):
+    def __init__(self, query_runner: QueryRunner, namespace: str, server_version: ServerVersion) -> None:
+        if server_version < ServerVersion(2, 4, 0):
+            raise ValueError("The new Cypher projection is only supported since GDS 2.4.0.")
+        super().__init__(query_runner, namespace, server_version)
+
+    def project(
+        self,
+        graph_name: str,
+        query: str,
+        params: Optional[Dict[str, Any]] = None,
+        database: Optional[str] = None,
+    ) -> Tuple[Graph, "Series[Any]"]:
+        """
+        Run a Cypher projection.
+        The provided query must end with a `RETURN gds.graph.project(...)` call.
+
+        Parameters
+        ----------
+        graph_name: str
+            the name of the graph to project
+        query: str
+            the Cypher projection query
+        params: Dict[str, Any]
+            parameters to the query
+        database: str
+            the database on which to run the query
+
+        Returns
+        -------
+        A tuple of the projected graph and statistics about the projection
+        """
+
+        graph_name_param = self._find_return_clause_graph_name(query)
+
+        if not graph_name_param.startswith("$"):
+            raise ValueError(
+                f"Invalid query, the `graph_name` must use a query parameter, but got `{graph_name_param}`: {query}"
+            )
+
+        graph_name_param = graph_name_param[1:]
+
+        if params is not None and graph_name_param in params:
+            query_graph_name = params[graph_name_param]
+            if query_graph_name != graph_name:
+                raise ValueError(
+                    f"Invalid query, the `{graph_name_param}` parameter must be bound to `{graph_name}`: {query}"
+                )
+        else:
+            if params is None:
+                params = {}
+
+            params[graph_name_param] = graph_name
+
+        # See run_cypher
+        qr = self._query_runner
+
+        # The Arrow query runner should not be used to execute arbitrary Cypher
+        if isinstance(qr, ArrowQueryRunner):
+            qr = qr.fallback_query_runner()
+
+        result = qr.run_query(query, params, database, False)
+        result = result.squeeze()
+
+        return Graph(graph_name, self._query_runner, self._server_version), result  # type: ignore
+
+    __separators = re.compile(r"[,(.]")
+
+    def _find_return_clause_graph_name(self, query: str) -> str:
```

Comment:
Should this maybe live in some util file? With separate tests .. :)

Reply:
Dunno about a util file, it is not really generic in what it does. I did add some tests though.